desktop/src/main/kotlin/io/spine/examples/pingh/desktop/Login.kt

```diff
@@ -0,0 +1,328 @@
+ */
+@Composable
+private fun ErrorMesage(isShown: Boolean) {
+    if (isShown) {
+        Text(
+            text = "Must consist of alphanumeric characters and dashes, " +
+                    "without consecutive dashes or dashes at the beginning or end.",
+            modifier = Modifier
+                .width(155.dp)
+                .height(30.dp)
+                .absoluteOffset(x = 15.dp, y = 42.dp),
+            color = MaterialTheme.colorScheme.error,
+            style = MaterialTheme.typography.bodySmall
+        )
+    }
+}
+
+/**
+ * Displays a button on the login form.
+ */
+@Composable
+private fun LogInButton(
+    enabled: Boolean,
+    onClick: () -> Unit
+) {
+    Button(
+        onClick = onClick,
+        modifier = Modifier
+            .width(120.dp)
+            .height(40.dp),
+        enabled = enabled,
+        colors = ButtonDefaults.buttonColors(
+            containerColor = MaterialTheme.colorScheme.primary,
+            contentColor = MaterialTheme.colorScheme.onPrimary
+        )
+    ) {
+        Text(
+            text = "Log in",
+            style = MaterialTheme.typography.displayMedium
+        )
+    }
+}
+
+/**
+ * Creates a new `Username` with the specified value without validation.
+ *
+ * The `Username` value cannot be empty. However, if the login field is empty,
+ * the `Username` will also be empty. To avoid exceptions during building,
+ * the `buildPartial` method is used, which does not check the `Username` value.
+ *
+ * Before sending this `Username` to the server, ensure to [validate] the value.
+ */
+private fun KClass<Username>.buildWithoutValidationBy(value: String) =
+    Username.newBuilder()
+        .setValue(value)
+        .buildPartial()
+
+/**
+ * Returns `true` if the `Username` is valid according to GitHub criteria; otherwise, returns `false`.
+ *
+ * A valid GitHub username must:
+ *
+ * - Consist of alphanumeric characters and dashes (`'-'`);
+ * - Not have consecutive dashes or dashes at the beginning or end;
+ * - Not exceed 39 characters.
+ *
+ * @see <a href="https://docs.github.com/en/enterprise-server@3.9/admin/managing-iam/iam-configuration-reference/username-considerations-for-external-authentication">
+ *     Username considerations for external authentication</a>
+ */
+@Suppress("ReturnCount") // To preserve the integrity of the algorithm,
+// the number of `return` is exceeded.
+private fun Username.validate(): Boolean {
+    if (this.value.length !in 1..maxLengthOfUsername) {
+        return false
+    }
+    var previous = '-'
+    this.value.forEach { current ->
+        if (current.validateGiven(previous)) {
+            return false
+        }
+        previous = current
+    }
+    return previous != '-'
+}
+
+/**
+ * Returns `true` if this character is alphanumeric or
+ * if it is a dash and `previous` is not a dash. Otherwise, returns `false`.
+ */
+private fun Char.validateGiven(previous: Char): Boolean =
```

Comment:
Let's move general-type extensions to a separate file. They should be `internal` there.